### PR TITLE
reindex: automatically choose the number of slices

### PR DESCRIPTION
**Pull request by andyb-elastic**
For #24547

Add an option to reindex, update by query, and delete by query to set `slices` to `auto` rather than a specific number. The number of slices it chooses will be the lowest number of shards among the source indices, up to a constant ceiling. 

I chose the ceiling arbitrarily as 20 here. Next I want to do some of the rally benchmarking mentioned in the original issue to find what value makes the most sense.

In the interest of maintaining the original behavior, when slices is set to auto and there is only one shard, it will handle the request as if it was not sliced (i.e. `slices` defaults to 1). 

This also adds unit tests for these APIs with multiple source indices

This includes the language from #25582 so I'll close that when this is merged

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/reindex/AbstractBulkByScrollRequest.java [hunk 1]
```diff
@@ -409,9 +399,9 @@ public void readFrom(StreamInput in) throws IOException {
         maxRetries = in.readVInt();
         requestsPerSecond = in.readFloat();
         if (in.getVersion().onOrAfter(Version.V_5_1_1)) {
-            slices = in.readVInt();
+            slices = new Slices(in);
         } else {
-            slices = 1;
+            slices = Slices.DEFAULT;
         }
```
Comment (andy-elastic) [inline core/src/main/java/org/elasticsearch/index/reindex/AbstractBulkByScrollRequest.java]: I'm not sure how backwards compatibility should work here. This change uses `-1` as the value here for serializing the auto setting, which won't be a valid slices int value in earlier versions. Something like

```
readFrom {
  if version on or after 5.1.1 {
    slices = new Slices(in)
  } else {
    slices = default
  }
}

writeTo {
  if version on or after 5.1.1 and before 6.1.0 {
    if slices is auto {
      throw exception
    } else {
      slices.writeTo(out)
    }
  } else if version on or after 6.1.0
    slices.writeTo(out)
  } else {
    if slices > 1 or slices is auto, throw exception
  }
}

File: core/src/main/java/org/elasticsearch/index/reindex/Slices.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+import org.elasticsearch.common.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Represents the setting for the number of slices used by a BulkByScrollRequest. Valid values are positive integers and "auto". The
+ * "auto" setting defers choosing the number of slices to the request handler.
+ */
+public final class Slices implements ToXContent, Writeable {
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/index/reindex/Slices.java]: Why do we need an entirely new class just to wrap a single integer? The boolean methods here could be static methods taking the int?

File: core/src/main/java/org/elasticsearch/index/reindex/AbstractBulkByScrollRequest.java [hunk 3]
```diff
@@ -152,7 +152,7 @@ public ActionRequestValidationException validate() {
                             + size + "]",
                     e);
         }
-        if (searchRequest.source().slice() != null && slices != 1) {
+        if (searchRequest.source().slice() != null && !slices.equals(Slices.DEFAULT)) {
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/index/reindex/AbstractBulkByScrollRequest.java]: We prefer to use `== false` for negation because it is easier to see visually. You will find this pattern in both Elasticsearch and Lucene.

**Review by rjernst — COMMENTED** [review]
This is not a full review, just a couple random things I saw while glancing through the change.

File: core/src/main/java/org/elasticsearch/index/reindex/AbstractBulkByScrollRequest.java (hunk 3 above)
Comment (andy-elastic) [inline core/src/main/java/org/elasticsearch/index/reindex/AbstractBulkByScrollRequest.java]: Agree, that's much clearer visually

File: core/src/main/java/org/elasticsearch/index/reindex/Slices.java (hunk 2 above)
Comment (andy-elastic) [inline core/src/main/java/org/elasticsearch/index/reindex/Slices.java]: My thinking here was that rather than duplicate this logic (is `slices` a number or "auto") everywhere that distinction is relevant, I'd encapsulate it into a class. That and I prefer to avoid magic numbers if possible because they're less obvious than using a separate type.

That said, I definitely understand why this may not be desirable

>The boolean methods here could be static methods taking the int?

Do you mean something like, treat slices=0 as auto and just have this method somewhere

```java
public static boolean isAuto(int slices) {
  if (slices == 0) {
    return true
  } else if (slices > 0) {
    return false;
  } else {
    throw new InvalidArgumentException();
  }
}
```

File: core/src/main/java/org/elasticsearch/index/reindex/Slices.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    public static final Slices ONE = of(1);
+    public static final Slices DEFAULT = ONE;
+
+    private final int count;
+
+    private Slices(int count) {
+        this.count = count;
+    }
+
+    public Slices(StreamInput stream) throws IOException {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/index/reindex/Slices.java]: I like to put the reading code right under the writing code so they are easy to eyeball together.

**Review by nik9000 — COMMENTED** [review]
I had a look and left a few comments but didn't read all the way through.

File: core/src/main/java/org/elasticsearch/index/reindex/Slices.java (hunk 2 above)
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/index/reindex/Slices.java]: Something like that. Although I'm not sure if that is even really necessary. Just have a constant `AUTO_SLICES = -1`. This is similar the `NO_DOC` constant in lucene for docid. And then check for equality with the constant.

Comment (rjernst) [inline core/src/main/java/org/elasticsearch/index/reindex/Slices.java]: But since you want the number to be positive, I would also use the value `0`, so there is no gap for validation.

File: core/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java [hunk 5]
```diff
@@ -43,32 +43,118 @@
...
-     * Overridden to force children to return compatible status.
+        parentWorker = new ParentBulkByScrollWorker(this, slices);
+    }
+
+    /**
+     * Returns the worker object that tracks the state of sliced subtasks. Throws IllegalStateException if this task is not set to be
+     * a parent task.
      */
-    public abstract BulkByScrollTask.Status getStatus();
+    public ParentBulkByScrollWorker getParentWorker() {
```
Comment (andy-elastic) [inline core/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java]: It looks like these methods have to be public to be accessible to the classes in the reindex module that use them. They get IllegalAccessError in integration even though they're in the same package. I'm not really sure how module loading works, but if it uses a different classloader than core I think it would disallow package-scope access in this case.

Sort of related: would it make sense to move some of the stuff covered in this PR that's in core to the reindex module?

**Comment by andy-elastic** [comment]
@rjernst @nik9000 thanks for the feedback, I think I got all the points covered here so far

File: core/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java (hunk 5 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java]: > I'm not really sure how module loading works, but if it uses a different classloader than core I think it would disallow package-scope access in this case.

Yeah. I think it is rude to have the same package in two modules....

>Sort of related: would it make sense to move some of the stuff covered in this PR that's in core to the reindex module?

I wouldn't worry too much about it for now. I believe @rjernst is working on something that'll let us more everything over to reindex's module again which'd be lovely.

**Review by nik9000 — CHANGES_REQUESTED** [review]
I left a bunch of minor stuff and some stylistic stuff. I think it quite close though.

File: core/src/main/java/org/elasticsearch/index/reindex/AbstractBulkByScrollRequest.java [hunk 6]
```diff
@@ -152,8 +156,8 @@ public ActionRequestValidationException validate() {
                             + size + "]",
                     e);
         }
-        if (searchRequest.source().slice() != null && slices != 1) {
-            e = addValidationError("can't specify both slice and workers", e);
+        if (searchRequest.source().slice() != null && slices != DEFAULT_SLICES) {
+            e = addValidationError("can't set a specific single slice for this request and multiple slices", e);
```
Comment (andy-elastic) [inline core/src/main/java/org/elasticsearch/index/reindex/AbstractBulkByScrollRequest.java]: Yeah I think that makes sense. When I wrote that I was thinking about the child tasks having a slice builder set in automatic slicing, but this aligns more with how the api is described

File: core/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java [hunk 7]
```diff
@@ -43,32 +43,127 @@
...
+    public void setSliceChild(float requestsPerSecond) {
+        setSliceChild(requestsPerSecond, null);
+    }
+
+    /**
+     * Sets this task to be a slice child task that performs search requests
+     * @param requestsPerSecond How many search requests per second this task should make
+     * @param sliceId If this is is a sliced task, which slice number this task corresponds to
+     */
+    public void setSliceChild(float requestsPerSecond, Integer sliceId) {
```
Comment (andy-elastic) [inline core/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java]: Cool, I'd seen that but wasn't sure if it was intended for use outside of dependency injection

File: core/src/main/java/org/elasticsearch/tasks/Task.java [hunk 8]
```diff
@@ -89,7 +89,7 @@ public final TaskInfo taskInfo(String localNodeId, boolean detailed) {
     /**
      * Build a proper {@link TaskInfo} for this task.
      */
-    protected final TaskInfo taskInfo(String localNodeId, String description, Status status) {
+    public final TaskInfo taskInfo(String localNodeId, String description, Status status) {
```
Comment (andy-elastic) [inline core/src/main/java/org/elasticsearch/tasks/Task.java]: Yeah, I didn't want to do this but couldn't think of a good way since the caller isn't part of that type hierarchy anymore. I'll look for a better alternative

File: core/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java [hunk 9]
```diff
@@ -43,32 +43,127 @@
...
+    public ParentBulkByScrollWorker getSlicesParentWorker() {
+        if (!isSlicesParent()) {
+            throw new IllegalStateException("This task is not set as a parent worker");
+        }
+        return parentWorker;
+    }
 
     /**
-     * Build the status for this task given a snapshot of the information of running slices.
+     * Returns true if this task is a child task that performs search requests. False otherwise
```
Comment (andy-elastic) [inline core/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java]: It'll return true for single sliced requests since they get set with a child worker, even if `slices` was `auto`. I agree the names aren't great and there's a lot of overlap, I'll see if I can find some better ones

File: docs/reference/docs/delete-by-query.asciidoc [hunk 10]
```diff
@@ -463,6 +472,11 @@ Which results in a sensible `total` like this one:
 ----------------------------------------------------------------
 // TESTRESPONSE
 
+Setting `slices` to `auto` will let Elasticsearch choose the number of slices
+to use. This setting will use one slice per shard, up to a certain limit. If
+there are multiple source indices, it will choose the number of slices based
```
Comment (andy-elastic) [inline docs/reference/docs/delete-by-query.asciidoc]: Sounds good, I'll open an issue for it after this is merged. I haven't looked at how search requests with multiple indices work so I'm not familiar with what would change in that case

File: core/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java (hunk 7 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java]: There are a couple of `@Nullable`s on the classpath. The one in `org.elasticsearch.common` is probably the right one.

**Comment by andy-elastic** [comment]
@nik9000 I changed Worker -> TaskState since that's mostly what it is, and Parent -> Leader and Child -> worker to make the relationship more clear. I think these names are a little better than what I had originally

For your comment about the [supplier in the parallelization helper](https://github.com/elastic/elasticsearch/pull/26030/files#diff-cf20cc1a937fbf676a470e736a73f875R69) can you elaborate on why a Runnable would be better? I just went with the stricter behavior because it seemed like starting an `AbstractAsyncBulkByScrollAction` was the goal here

**Comment by nik9000** [comment]
> For your comment about the supplier in the parallelization helper can you elaborate on why a Runnable would be better? I just went with the stricter behavior because it seemed like starting an AbstractAsyncBulkByScrollAction was the goal here

Sure! Reading the supplier made me think you were getting something. And then when I realized what you were getting I thought "but doesn't he need to customize that for every request?" and then I realized why you don't, it is because you only call it at all when starting the "working" request. You skip it when you are parallelizing. So then I thought "why not just call it something that has to do with running the request?" And then I thought, "why is it a supplier when it really is about running stuff?"

Basically I got confused about what it was for so I suggested renaming it. You don't have to make it into a `Runnable` but it'd be nice to rename it so it is obvious that it is only for the working portion.

File: core/src/main/java/org/elasticsearch/index/reindex/AbstractBulkByScrollRequest.java [hunk 11]
```diff
@@ -345,6 +345,9 @@ public boolean getShouldStoreResult() {
      * The number of slices this task should be divided into. Defaults to 1 meaning the task isn't sliced into subtasks.
      */
     public Self setSlices(int slices) {
+        if (slices < 0) {
+            throw new IllegalArgumentException("[slices] must be at least 0");
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/index/reindex/AbstractBulkByScrollRequest.java]: Just in case this is ever thrown it'd be nice to have the error message more like "slices must be at least 0 but was [$value]".

**Review by nik9000 — APPROVED** [review]

**Comment by andy-elastic** [comment]
I see what you mean, I think it's more clear if it's a runnable.

It looks like the last CI build failed because the node was low on resources, I'll let this finish before merging in the interest of getting a green build
